core/unpledge/unpledge.go

```diff
@@ -0,0 +1,305 @@
+package unpledge
+
+import (
+	"bufio"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/EnsurityTechnologies/logger"
+	"github.com/rubixchain/rubixgoplatform/core/wallet"
+
+	"golang.org/x/crypto/sha3"
+)
+
+type UnPledge struct {
+	log logger.Logger
+}
+
+func (un *UnPledge) GetPledgedTokenID(did string) ([]string, error) {
+	w := wallet.Wallet{}
+	var tokenIDs []string
+	pledgedTokens, err := w.GetAllPledgedTokens(did)
+	if err != nil {
+		return nil, err
+	}
+	for _, token := range pledgedTokens {
+		tokenIDs = append(tokenIDs, token.TokenID)
+	}
+	return tokenIDs, nil
+}
+
+func (un *UnPledge) GetLastBlockReceiverDID(TokenID string) (string, error) {
+	w := wallet.Wallet{}
+	lastBlock := w.GetLatestTokenBlock(TokenID)
+	if lastBlock == nil {
+		un.log.Error("latest block for token %s not found", TokenID)
+		return "", nil
+	}
+	ReceiverDID := lastBlock.GetReceiverDID()
+	return ReceiverDID, nil
+}
+
+func (un *UnPledge) GetLastBlockTransactionID(TokenID string) (string, error) {
+	w := wallet.Wallet{}
+	lastBlock := w.GetLatestTokenBlock(TokenID)
+	if lastBlock == nil {
+		un.log.Error("latest block for token %s not found", TokenID)
+		return "", nil
+	}
+	TransactionID := lastBlock.GetTid()
+	return TransactionID, nil
+}
+
+func (un *UnPledge) Concat(TokenID string, ReceiverDID string) (string, error) {
+	if TokenID == "" {
+		err := errors.New("TokenID cannot be empty")
+		un.log.Error(err.Error())
+		return "", err
+	}
+	if ReceiverDID == "" {
+		err := errors.New("ReceiverDID cannot be empty")
+		un.log.Error(err.Error())
+		return "", err
+	}
+	valueConcated := TokenID + ReceiverDID
+	return valueConcated, nil
+}
+
+func (un *UnPledge) CalcSHA_256Hash(input string) string {
+	hash := sha256.Sum256([]byte(input))
+	hashedInput := fmt.Sprintf("%x", hash)
+	return hashedInput
+}
+
+func (un *UnPledge) CalcSHA_3_256Hash(input string) string {
+	hash := sha3.Sum256([]byte(input))
+	hashedInput := fmt.Sprintf("%x", hash)
+	return hashedInput
+}
+
+func (un *UnPledge) CalcSHA3_256Hash1000Times(input string) string {
+	var hashedInput string
+	for i := 0; i < 1000; i++ {
+		hash := sha3.Sum256([]byte(input))
+		hashedInput = fmt.Sprintf("%x", hash)
+		input = hashedInput
+	}
+	return hashedInput
+}
+
+func (un *UnPledge) getCurrentLevel() int {
+	url := "http://13.76.134.226:9090/getCurrentLevel"
```

Comment:
Nothing to fix in the PR right away but we need to consider have a domain name for this, right? Will be a problem if we have to move away from machine that is hosting this server.